rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,226 @@
+...
+```
+**Note**: For `Merge on Read` tables, the log files will also go through the same hashing process and may not end up under
+the same prefix as the base parquet file for the FileSlice to which it belongs.
+
+The original table path will continue to store the `metadata folder` and `partition metadata` files:
+```
+s3://<table_bucket>/<hudi_table_name>/.hoodie/...
+s3://<table_bucket>/<hudi_table_name>/country=usa/.hoodie_partition_metadata
+s3://<table_bucket>/<hudi_table_name>/country=india/.hoodie_partition_metadata
+s3://<table_bucket>/<hudi_table_name>/country=uk/.hoodie_partition_metadata
+...
+```
+
+#### Hashing
+
+#####Option 1:
+We can re-use the implementations is `HashID` class to generate hash on `File Name` or `Partition + File Name`, which
+uses XX hash function with 32/64 bits (known for being fast).
+
+#####Option 2:
+To generate the prefixes we can use `Murmur 32 bit` hash, which is known for being fast and provides good distribution
+guarantees. We might have to further do bucketing and re-hash it to reduce the number of possible hashes from 2^32 to a
+slightly lower number, as it may be overkill to have that many unique hashes, which might result in scenarios
+where each file is under a different prefix.
+
+The hashing function should be made user configurable.
+
+### Maintain mapping to files
+
+In [RFC-15](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=147427331), we introduced an internal
+Metadata Table with a `files` partition that maintains mapping from partitions to list of files in the partition stored
+under `Table Path`. This mapping is kept up to date, as operations are performed on the original table. We will leverage
+the same to now maintain mappings to files stored at `Table Storage Path` under different prefixes.
+
+Here are some of the design considerations:
+
+1. Metadata table is a pre-requisite for federated storage to work. Since Hudi 0.11 we have enabled metadata table by
+default and hence this feature can be enabled by the users as long as they are not explicitly turning off metadata
+table, in which case we should throw an exception.
+
+2. The federated storage cannot be enabled on an existing table that is already bootstrapped with Hive storage
+layout. To switch to federated storage, the table will need to be re-bootstrapped with the new layout.
+
+3. The Instant metadata (`HoodieCommitMetadata`,`HoodieCleanMetadata` etc.) will always act as the source of file listing
+for metadata table to be populated.
+
+4. `HoodieCommitMetadata` currently stores `file name` instead of complete `file path`. We will have to modify commit
+metadata to store the complete file path instead of just file name, as the files are now distributed across several random
+prefix paths instead of a derivable table/partition path.
+
+5. If there is an error reading from Metadata table, we will not fall back listing from file system.
```

Comment:
How this would be possible w/o File System supporting Federated Storage (providing Hudi's components w/ a virtual view of the table, see my comment above for more context)?

Reply:
Having the same question as @alexeykudinkin here.  Does this mean that error reading metadata table completely blocks the ingestion if the prefix-based storage layout is enabled?

Reply:
If so, the repair tool is a must.